controlplane/kubeadm/internal/workload_cluster_coredns_test.go

```diff
@@ -287,62 +291,75 @@ kind: ClusterConfiguration
 	// Namespace/Name (e.g. The CoderDNS ConfigMap & Deployment, the kubeadm ConfigMap), it is not possible to run the use cases in parallel.
 	for _, tt := range tests {
 		g := NewWithT(t)
-		t.Log(tt.name)
-
-		for _, o := range tt.objs {
-			// NB. deep copy test object so changes applied during a test does not affect other tests.
-			o := o.DeepCopyObject().(client.Object)
-			g.Expect(testEnv.Create(ctx, o)).To(Succeed())
-		}
 
-		w := &Workload{
-			Client:          testEnv.GetClient(),
-			CoreDNSMigrator: tt.migrator,
-		}
-		err := w.UpdateCoreDNS(ctx, tt.kcp)
-		if tt.expectErr {
-			g.Expect(err).To(HaveOccurred())
-			return
-		}
-		g.Expect(err).ToNot(HaveOccurred())
+		t.Run(tt.name, func(t *testing.T) {
+			for _, o := range tt.objs {
+				// NB. deep copy test object so changes applied during a test does not affect other tests.
+				o := o.DeepCopyObject().(client.Object)
+				g.Expect(testEnv.Create(ctx, o)).To(Succeed())
+				// this makes sure that the cache is updated with the object
+				// to avoid 404 errors leading to test flakes
+				g.Eventually(func() bool {
+					key, _ := client.ObjectKeyFromObject(o)
+					err := testEnv.Get(ctx, key, o)
+					if err != nil && apierrors.IsNotFound(err) {
+						return false
+					}
+					return true
+				}, "10s").Should(BeTrue())
+			}
 
-		// Assert that CoreDNS updates have been made
-		if tt.expectUpdates {
-			// assert kubeadmConfigMap
-			var expectedKubeadmConfigMap corev1.ConfigMap
-			g.Expect(testEnv.Get(ctx, ctrlclient.ObjectKey{Name: kubeadmConfigKey, Namespace: metav1.NamespaceSystem}, &expectedKubeadmConfigMap)).To(Succeed())
-			g.Expect(expectedKubeadmConfigMap.Data).To(HaveKeyWithValue("ClusterConfiguration", ContainSubstring("1.7.2")))
-			g.Expect(expectedKubeadmConfigMap.Data).To(HaveKeyWithValue("ClusterConfiguration", ContainSubstring("k8s.gcr.io/some-repo")))
+			// Register cleanup function
+			t.Cleanup(func() {
+				// Cleanup test objects (and wait for deletion to complete).
+				_ = testEnv.Cleanup(ctx, tt.objs...)
+				g.Eventually(func() bool {
+					for _, o := range []client.Object{cm, depl, kubeadmCM, badCM} {
+						key, _ := client.ObjectKeyFromObject(o)
```

Comment:
I believe you need to restore `o := o.DeepCopyObject().(client.Object)` here. Otherwise, it modifies the variables (including setting the ResourceVersion), and you won't be able to perform subsequent creates.

Reply:
Oh I see why this might be a problem now, after you pointed this out. The reference `o` might be getting overridden with the fetched object.